src/psyclone/psyir/tools/definition_use_chains.py

```diff
@@ -0,0 +1,629 @@
+                                Routine exists.
+    :type control_flow_region: None or
+                               List[:py:class:`psyclone.psyir.nodes.Node`]
+    :param bool is_local: Optional argument to define whether reference is a
+                          local variable or not. Default behaviour is to check
+                          if its interface is an AutomaticInterface or not.
+    :param int start_point: Optional argument to define a start point for the
+                            dependency search.
+    :param int stop_point: Optional argument to define a stop point for the
+                           dependency search.
+
+    :raises TypeError: If one of the arguments is the wrong type.
+
+    """
+
+    def __init__(
+        self,
+        reference,
+        control_flow_region=None,
+        is_local=None,
+        start_point=None,
+        stop_point=None,
+    ):
+        if not isinstance(reference, Reference):
+            raise TypeError(f"The reference passed into a DefinitionUseChain "
+                            f"must be a Reference but found "
+                            f"'{type(reference).__name__}'.")
+        self._reference = reference
+        # Store the absolute position for later.
+        self._reference_abs_pos = reference.abs_position
+        # To enable loops to work correctly we can set the start/stop point
+        # and not just use base it on the reference's absolute position
+        if start_point and not isinstance(start_point, int):
+            raise TypeError(f"The start_point passed into a "
+                            f"DefinitionUseChain must be an int but found "
+                            f"'{type(start_point).__name__}'.")
+        if stop_point and not isinstance(stop_point, int):
+            raise TypeError(f"The stop_point passed into a "
+                            f"DefinitionUseChain must be an int but found "
+                            f"'{type(stop_point).__name__}'.")
+        self._start_point = start_point
+        self._stop_point = stop_point
+        if control_flow_region is None:
+            self._scope = [reference.ancestor(Routine)]
+            if self._scope[0] is None:
+                self._scope = reference.root.children[:]
+        else:
+            # We need a list of regions for control flow.
+            if not isinstance(control_flow_region, list):
+                raise TypeError(
+                    f"The control_flow_region passed into a "
+                    f"DefinitionUseChain must be a list but found "
+                    f"'{type(control_flow_region).__name__}'."
+                )
+            self._scope = control_flow_region
+        if is_local is None:
+            # TODO Work out if symbol is local to the scope or
+            # globally accessible. Anything that isn't a local
+            # variable in a Routine has to be assumed as globally
+            # accessible for now.
```

Comment:
Yeah - the TODO is out of date.

Since here it plays very safe I could see a use case where some code might know whether an argument is actually local or not.

Reply:
I prefer to remove it for now for simplify slightly the implementation (both the argument and the private variable - as I see it is only used once?) Also note that symbols have properties to avoid using the isinstance pattern for these, e.g. `symbol.is_automatic`